rfcs/2021-06-15-shipping-blade.md

```diff
@@ -0,0 +1,355 @@
+    // this is allowed too
+    import Button from '@razorpay/blade/_SomeInternalButton'
+    ```
+3. As library authors we are not responsible for tree shaking, we depend on the consumer applications to take care of that. We'll just give them transpiled code so their app tools can understand.
+4. Working of library totally depends on the tools config and different versions of consumer apps. It should at least have the configs(babel presets/plugins) in which the library is built.
+> 📝 We can create our custom plugins and ask consumers to import it but that's again too much effort to maintain another piece.
+
+### Shall we Bundle(using rollup/similar tools)?
+It'll bundle the code into one single file/multiple(depending on our config).Before we take a call let's see a checklist of what transpiling can/can't do.
+1. We can use library with the script tag in html(umd bundle).
+    ```html
+    <script src="https://unpkg.com/@razorpay/blade@0.1.1/index.js" crossorigin></script>
+    ```
+2. The imports are restricted to what we re-export and bundle.
+    ```js
+    // this is allowed
+    import { Button } from '@razorpay/blade'
+    // this is not allowed as it's not re-exported
+    import Button from '@razorpay/blade/_SomeInternalButton'
+    ```
+3. As library authors we are responsible for tree shaking since we'll be bundling the code which will be ready to be used by the consumers.
+4. Since we ship the packaged code we don't care what tools config or versions the consumer app is using. We'll ship code that they just need to plug in and it would work.
+
+### Conclusion
+I feel we below is the checklist that we need.
+
+✅ Can use library with the script tag in html(umd bundle).
+
+✅ Library author responsible for tree shaking
+
+❌ Import anything from anywhere
+
+❌ Working of library depends on the build tools config of consumer apps
+
+And this can be achieved by going with the approach of bundling using rollup.
+
+> Not choosing parcel for following reasons:
+> 1. Not familiar with the tool
+> 2. We need to bundle react-native as well so I'm not too sure how smooth it will be with parcel.
+
+## What module system to target?
+
+### CommonJS(a.k.a cjs)
+* Everything that we see using `require` or `module.exports` is commonjs
+* Not great at tree shaking capabilities because when you import you get an object.
+* More code shipped to the browsers which means more download, parse execute times.
+### ES Modules(a.k.a esm)
+* Everything that we see using `import` or `export` is esm.
+* It's great at tree shaking as you can determine imports and exports at compile time (statically) — the tools only have to look at the source code, they don’t have to execute it.
+* Less code shipped to the browsers which means better download, parse execute times.
+
+### Conclusion
+We should definitely going with ESM. There might be few gotchas and unknowns that might pop up but we'll move ahead and fix them instead of taking a back seat and ship CJS.
```

Comment:
```suggestion
We should definitely go with ESM. There might be few gotchas and unknowns that might pop up but we'll move ahead and fix them instead of taking a back seat and ship CJS.
```